File: LEETCODE-207-course-schedule.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int,vector<int>> adj;
        for(vector<int>& task: prerequisites) {
            adj[task[0]].push_back(task[1]);
        }
        unordered_map<int,int> indegree;
        for(int course=0; course<numCourses; course++) {
            for(int& it: adj[course]) {
                indegree[it]++;
            }
        }
        queue<int> q;
        for(int course=0; course<numCourses; course++) {
            if(indegree[course] == 0) {
                q.push(course);
            }
        }
        vector<int> topo;
        while(!q.empty()) {
            topo.push_back(q.front());
            for(int& neighbour: adj[q.front()]) {
                indegree[neighbour]--;
                if(indegree[neighbour] == 0) {
                    q.push(neighbour);
                }
            }
            q.pop();
        }
        return topo.size() == numCourses;
    }
};

File: LEETCODE_207_Course_Schedule.java
import java.util.*;

public class LEETCODE_207_Course_Schedule {

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] indegree = new int[numCourses];
        Arrays.fill(indegree, 0);
        for(int[] p: prerequisites) {
            int course = p[0], prerequisite = p[1];
            if(graph.containsKey(prerequisite) == false) {
                graph.put(prerequisite, new ArrayList<>());
            }
            graph.get(prerequisite).add(course);
            indegree[course]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for(int node = 0; node < numCourses; node++) {
            if(indegree[node] == 0) {
                q.offer(node);
            }
        }
        List<Integer> toposort = new ArrayList<>();
        while(q.isEmpty() == false) {
            int node = q.poll();
            toposort.add(node);
            if(graph.containsKey(node) == true) {
                for(int neighbour: graph.get(node)) {
                    indegree[neighbour]--;
                    if(indegree[neighbour] == 0) {
                        q.offer(neighbour);
                    }
                }
            }
        }
        return toposort.size() == numCourses;
    }
}